File: WebAPI/WebAPI.Application/Services/MenuServices/MenuCategoryService/MenuCategoryService.cs
using AutoMapper;
using Microsoft.EntityFrameworkCore;
using WebAPI.Application.DTO.Element;
using WebAPI.Application.DTO.MenuDTO.MenuCategory;
using WebAPI.Application.DTO.Other;
using WebAPI.Domain.Entities;
using WebAPI.Domain.Entities.Menu;
using WebAPI.Repository.Data;

namespace WebAPI.Application.Services.MenuServices.MenuCategoryService;

public class MenuCategoryService(AppDbContext context, IMapper mapper) : IMenuCategoryService
{
    public async Task<List<MenuCategoryDto>> ReplaceAllByMenuIdAsync(int menuId, int userId, List<CreateMenuCategoryDto> createMenuCategoryDtos)
    {
        var oldCategories = await context.MenuCategories.Where(e => e.UserId == userId && e.MenuId == menuId).ToListAsync();
        oldCategories.ForEach(e => { context.Remove(e); });
        await context.SaveChangesAsync();
        var menuCategories = createMenuCategoryDtos.Select(mapper.Map<MenuCategory>).ToList();
        menuCategories.ForEach(e =>
        {
            e.UserId = userId;
            context.Add(e);
        });
        await context.SaveChangesAsync();
        return menuCategories.Select(mapper.Map<MenuCategoryDto>).ToList();
    }

    public async Task<MenuCategoryDto> CreateAsync(int userId, CreateMenuCategoryDto createMenuCategoryDto)
    {
        var menuCategory =
            await context.CreateFromDtoWithUserIdAsync<MenuCategory, CreateMenuCategoryDto>(userId,
                createMenuCategoryDto, mapper);
        return mapper.Map<MenuCategoryDto>(menuCategory);
    }

    public async Task DeleteByIdAsync(int id, string menuType, int userId)
    {
        await context.DeleteByIdWithUserIdAsync<MenuCategory>(id, userId);
        var cats = await context.MenuCategories
            .Where(e => e.UserId == userId && e.MenuType == menuType)
            .OrderBy(e => e.Position)
            .ToListAsync();
        for (var i = 0; i < cats.Count; i++)
            cats[i].Position = i + 1;
        await context.SaveChangesAsync();
    }

    public async Task UpdateByIdAsync(int id, int userId, UpdateMenuCategoryDto updateMenuCategoryDto)
    {
        await context.UpdateFromDtoWithUserIdAsync<MenuCategory, UpdateMenuCategoryDto>(id, userId,
            updateMenuCategoryDto, mapper);
    }

    public async Task UpdatePositionsByMenuIdAsync(int menuId, string menuType, int userId, List<PositionDto> categoryPositions)
    {
        var categories = await context.MenuCategories.Where(e => e.UserId == userId && e.MenuId == menuId && e.MenuType == menuType).ToListAsync();
        if (categoryPositions.Count != categories.Count)
            
            throw new Exception("More/Less positions than elements");
        foreach (var categoryPosition in categoryPositions)
        {
            var category = categories.FirstOrDefault(e => e.Id == categoryPosition.Id);
            if (category != null)
                category.Position = categoryPosition.Position;
            else
                throw new Exception("Position with Id to an element that doesn't exist");
        }

        await context.SaveChangesAsync();
    }

    public async Task UpdateVisibilityByIdAsync(int id, int userId, UpdateMenuCategoryVisibilityDto updateVisibilityDto)
    {
        await context.UpdateFromDtoWithUserIdAsync<MenuCategory, UpdateMenuCategoryVisibilityDto>(id, userId,
            updateVisibilityDto, mapper);
    }
}

File: WebApp/src/app/core/http/services/menu-services/menu-category/menu-category.service.ts
import {inject, Injectable} from '@angular/core';
import {CreateMenuCategoryDto} from "../../../dto/menu-dto/menu-category/create-menu-category.dto";
import {HttpClient} from "@angular/common/http";
import {EnvironmentService} from "../../../../services/environment/environment.service";
import {MenuCategoryDto} from "../../../dto/menu-dto/menu-category/menu-category.dto";
import {UpdateMenuCategoryDto} from "../../../dto/menu-dto/menu-category/update-menu-category.dto";
import {UpdateMenuCategoryVisibilityDto} from "../../../dto/menu-dto/menu-category/update-menu-category-visibility.dto";
import {PositionDto} from "../../../dto/other/position.dto";

@Injectable({
  providedIn: 'root',
})
export class MenuCategoryService {
  private readonly http = inject(HttpClient)
  private readonly environment = inject(EnvironmentService)
  private readonly baseUrl

  constructor() {
    this.baseUrl = this.environment.getBaseUrl() + '/MenuCategory'
  }

  public create(createMenuCategoryDto: CreateMenuCategoryDto) {
    return this.http.post<MenuCategoryDto>(
      `${this.baseUrl}/Create`,
      createMenuCategoryDto,
    )
  }

  public updateById(id: number, updateMenuCategoryDto: UpdateMenuCategoryDto) {
    return this.http.patch(
      `${this.baseUrl}/UpdateById/${id}`,
      updateMenuCategoryDto,
    )
  }

  public updateVisibilityById(id: number, updateVisibilityDto: UpdateMenuCategoryVisibilityDto) {
    return this.http.patch(
      `${this.baseUrl}/UpdateVisibilityById/${id}`,
      updateVisibilityDto,
    )
  }

  public updatePositionsByMenuId(menuId: number, menuType: string, categoryPositions: PositionDto[]) {
    return this.http.patch(
      `${this.baseUrl}/UpdatePositionsByMenuId/${menuId}/${menuType}`,
      categoryPositions
    )
  }

  public replaceAllByMenuId(menuId: number, createMenuCategoryDtos: CreateMenuCategoryDto[]) {
    return this.http.put(
      `${this.baseUrl}/ReplaceAllByMenuId/${menuId}`,
      createMenuCategoryDtos
    )
  }

  public deleteById(id: number, menuType: string) {
    return this.http.delete(`${this.baseUrl}/DeleteById/${menuType}/${id}`)
  }
}
